subprocess.hpp

```diff
@@ -136,69 +82,28 @@ std::future<int> async(const std::string commandPath, const std::vector<std::str
  * This delay must exist, as several programs may not output a line for each line input.
  *  Consider grep - it will not output a line if no match is made for that input. */
 class ProcessStream {
-    int statusCode;
-    pid_t childPid;
-    int inpipefd[2];
-    int outpipefd[2];
-    FILE* childStdout;
-
-    public:
-    ProcessStream(const std::string& commandPath, 
-            const std::vector<std::string>& commandArgs,
-            std::list<std::string>& stringInput) {
-        // based off https://stackoverflow.com/a/6172578
-        childPid = 0;
-
-        // construct the argument list (unfortunately, the C api wasn't defined with C++ in mind, so we have to abuse const_cast)
-        // see: https://stackoverflow.com/a/190208
-        std::vector<char*> cargs;
-        // the process name must be first for execv
-        cargs.push_back(const_cast<char*>(commandPath.c_str()));
-        for (const std::string& arg : commandArgs) {
-            cargs.push_back(const_cast<char*>(arg.c_str()));
-        }
-        // must be terminated with a nullptr for execv
-        cargs.push_back(nullptr);
-
-        pipe(inpipefd);
-        pipe(outpipefd);
-        childPid = fork();
-        // child
-        if (childPid == 0) {
-            dup2(outpipefd[0], STDIN_FILENO);
-            dup2(inpipefd[1], STDOUT_FILENO);
-            dup2(inpipefd[1], STDERR_FILENO);
-
-            // XXX: test (close the stdin..?)
-            close(outpipefd[1]);
-
-            //ask kernel to deliver SIGTERM in case the parent dies
-            prctl(PR_SET_PDEATHSIG, SIGTERM);
-
-            execv(commandPath.c_str(), cargs.data());
-            // Nothing below this line should be executed by child process. If so, 
-            // it means that the execl function wasn't successfull, so lets exit:
-            exit(1);
-        }
-
-        close(outpipefd[0]);
-        close(inpipefd[1]);
-
-        childStdout = fdopen(inpipefd[0], "r");
-
+    Process childProcess;
+
+public:
+    ProcessStream(const std::string& commandPath,
+        const std::vector<std::string>& commandArgs,
+        std::list<std::string>& stringInput) {
+            
+        childProcess.start(commandPath,commandArgs);
+        
         // while our string queue is working, 
         while (!stringInput.empty()) {
             // write our input to the process's stdin pipe
             std::string newInput = stringInput.front();
             stringInput.pop_front();
-            write(outpipefd[1], newInput.c_str(), newInput.size());
+            childProcess.pipe.writeP(newInput);
```

Comment:
any reason why we have direct access to the pipe to write, rather than making this a member function? I feel like this breaks the whole concept of encapsulation

Reply:
Yeah seeing this after my comment above. I agree.